crates/bifrost/src/providers/replicated_loglet/rpc_routers.rs

```diff
@@ -55,15 +72,607 @@ impl LogServersRpc {
+
+        Ok(())
+    }
+}
+
+//todo: inflight remote append
+pub(crate) struct RemoteInflightAppend {
+    rpc_token: RpcToken<Appended>,
+    commit_resolver: LogletCommitResolver,
+    permit: OwnedSemaphorePermit,
+}
+
+#[derive(thiserror::Error, Debug, Clone)]
+#[error("error status received: {status:?}")]
+pub struct AppendStatusError {
+    pub status: Status,
+}
+
+impl MaybeRetryableError for AppendStatusError {
+    fn retryable(&self) -> bool {
+        !matches!(
+            self.status,
+            Status::Disabled
+                | Status::Malformed
+                | Status::OutOfBounds
+                | Status::SequencerMismatch
+                | Status::Sealed
+                | Status::Sealing
+        )
+    }
+}
+
+pub enum BreakReason {
+    ConnectionClosed,
+    AppendError(AppendError),
+    Shutdown(ShutdownError),
+}
+
+/// Maintains a single connection to remote sequencer.
+///
+/// If the underlying connection failed, all pending commits
+/// on the same stream are immediately cancelled with an error.
+///
+/// It's up to the enqueuer to retry if the error is retryable.
+#[derive(Clone)]
+struct RemoteSequencerConnectionManager {
+    append: RpcRouter<Append>,
+    connection: Arc<Mutex<Option<RemoteSequencerConnection>>>,
+}
+
+impl RemoteSequencerConnectionManager {
+    pub fn new(rpc_router: RpcRouter<Append>) -> Self {
+        Self {
+            append: rpc_router,
+            connection: Arc::default(),
+        }
+    }
+
+    /// Gets or starts a new remote sequencer connection
+    async fn get<T: TransportConnect>(
+        &self,
+        networking: &Networking<T>,
+    ) -> Result<RemoteSequencerConnection, NetworkError> {
+        let mut guard = self.connection.lock().await;
+        if let Some(stream) = guard.deref() {
+            return Ok(stream.clone());
+        }
+
+        let connection = networking
+            .node_connection(networking.metadata().my_node_id().into())
+            .await?;
+        let stream = RemoteSequencerConnection::start(self.append.clone(), connection)?;
+
+        *guard = Some(stream.clone());
+
+        Ok(stream)
+    }
+
+    /// Renew a connection to a remote sequencer. This grantees that only a single connection
+    /// to the sequencer is available.
+    async fn renew<T: TransportConnect>(
+        &self,
+        networking: &Networking<T>,
+        old: &RemoteSequencerConnection,
+    ) -> Result<RemoteSequencerConnection, NetworkError> {
+        let mut guard = self.connection.lock().await;
+        let current = guard.as_ref().expect("connection has been initialized");
+
+        // stream has already been renewed
+        if old.connection != current.connection {
+            return Ok(current.clone());
+        }
+
+        let connection = networking
+            .node_connection(networking.metadata().my_node_id().into())
+            .await?;
+
+        let stream = RemoteSequencerConnection::start(self.append.clone(), connection)?;
```

Comment:
I wonder if we should make sure that the previous task is aborted in this case.

Reply:
Yeah, I see what you mean